src/objectid.ts

```diff
@@ -272,18 +292,19 @@ export class ObjectId extends BSONValue {
 
   /** @internal */
   serializeInto(uint8array: Uint8Array, index: number): 12 {
-    uint8array[index] = this.buffer[0];
-    uint8array[index + 1] = this.buffer[1];
-    uint8array[index + 2] = this.buffer[2];
-    uint8array[index + 3] = this.buffer[3];
-    uint8array[index + 4] = this.buffer[4];
-    uint8array[index + 5] = this.buffer[5];
-    uint8array[index + 6] = this.buffer[6];
-    uint8array[index + 7] = this.buffer[7];
-    uint8array[index + 8] = this.buffer[8];
-    uint8array[index + 9] = this.buffer[9];
-    uint8array[index + 10] = this.buffer[10];
-    uint8array[index + 11] = this.buffer[11];
+    const buffer = this.buffer || ByteUtils.fromHex(this.__id);
+    uint8array[index] = buffer[0];
+    uint8array[index + 1] = buffer[1];
+    uint8array[index + 2] = buffer[2];
+    uint8array[index + 3] = buffer[3];
+    uint8array[index + 4] = buffer[4];
+    uint8array[index + 5] = buffer[5];
+    uint8array[index + 6] = buffer[6];
+    uint8array[index + 7] = buffer[7];
+    uint8array[index + 8] = buffer[8];
+    uint8array[index + 9] = buffer[9];
+    uint8array[index + 10] = buffer[10];
+    uint8array[index + 11] = buffer[11];
     return 12;
```

Comment:
Maybe converting the pieces of the string one by one to byte will be more performant, something like: `parseInt(hex.slice(0, 2), 16)`.

In this case, we don't need to allocate a new buffer.

Reply:
I tried but with no luck. Seems to be a good deal slower using `parseInt(hex.slice(0, 2), 16)`

<img width="404" alt="image" src="https://github.com/mongodb/js-bson/assets/2258445/9464f7aa-8b92-4144-ba61-0e01f0809222">

Did some digging and I believe this is because NodeJS uses this native C++ module for hex encoding/decoding: [simdutf](https://github.com/simdutf/simdutf/?tab=readme-ov-file#how-fast-is-it) 

It's quite fast but there's probably some overhead in crossing the native boundary this frequently for such small strings. 

I'm excited for this proposal to land: https://github.com/tc39/proposal-arraybuffer-base64

Hopefully this means we could switch to UInt8Array everywhere and have performant hex encoding/decoding right in V8.

Reply:
Can you sent me the benchmark code? I want to try

Reply:
@H4ad Yeah for sure. I pushed the my benchmarks to this repo so you can run them all: https://github.com/SeanReece/js-bson-bench

I couldn't get the official bson-bench running so I hacked together some benchmarks myself.

The buffer.from vs parseInt vs buffer.write benchmark in question is the first one in the suite and it looks something like this

```ts
const buffer = Buffer.allocUnsafe(1000);
const hex = "48656c6c6f20576f726c64";
suite
.add("buffer.from + copy", function () {
    const buff = Buffer.from(hex, "hex");
    for (let i = 0; i < 12; i++) {
      buffer[i] = buff[i];
    }
  })
  .add("parseInt", function () {
    for (let i = 0; i < 12; i++) {
      buffer[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }
    return buffer;
  })
  // Test buffer.write vs buffer.from to copy hex string to buffer
  .add("buffer.write", function () {
    return buffer.write(hex, 0, 12, "hex");
  })
```

```bash
buffer.from + copy x 7,224,963 ops/sec ±2.57% (93 runs sampled)
parseInt x 2,984,423 ops/sec ±2.03% (89 runs sampled)
(-58.69%)
buffer.from + copy x 7,129,965 ops/sec ±3.50% (88 runs sampled)
buffer.write x 10,348,210 ops/sec ±0.96% (96 runs sampled)
(45.14%)
```

Reply:
The following code will run at ~10,6M:

```js
let temp = parseInt(hex.slice(0, 8), 16);

buffer[3] = temp & 0xff;
buffer[2] = (temp >> 8) & 0xff;
buffer[1] = (temp >> 16) & 0xff;
buffer[0] = (temp >> 24) & 0xff;

temp = parseInt(hex.slice(8, 16), 16);

buffer[7] = temp & 0xff;
buffer[6] = (temp >> 8) & 0xff;
buffer[5] = (temp >> 16) & 0xff;
buffer[4] = (temp >> 24) & 0xff;

temp = parseInt(hex.slice(16, 22), 16);

buffer[10] = temp & 0xff;
buffer[9] = (temp >> 8) & 0xff;
buffer[8] = (temp >> 16) & 0xff;
```

Reply:
@H4ad I think you missed that the random portion is 5 bytes, so to perform any bitwise operations on it we need to use BigInt, since integer bitwise is limited to 32-bit. That brings down performance :( 

This is how I was able to get it working:

```ts
    let temp = parseInt(hex.slice(0, 8), 16);

    buffer[3] = temp & 0xff;
    buffer[2] = (temp >> 8) & 0xff;
    buffer[1] = (temp >> 16) & 0xff;
    buffer[0] = (temp >> 24) & 0xff;

    temp = parseInt(hex.slice(8, 18), 16);

    buffer[8] = temp & 0xff;
    buffer[7] = (temp >> 8) & 0xff;
    buffer[6] = (temp >> 16) & 0xff;
    buffer[5] = (temp >> 24) & 0xff;
    buffer[4] = Number((BigInt(temp) >> 32n) & 0xffn);

    temp = parseInt(hex.slice(18, 24), 16);

    buffer[11] = temp & 0xff;
    buffer[10] = (temp >> 8) & 0xff;
    buffer[8] = (temp >> 16) & 0xff;
```

Now it's ~25% slower than Buffer.write.

<img width="413" alt="image" src="https://github.com/mongodb/js-bson/assets/2258445/5d5e67f0-87e5-45c8-ba53-8ff3404a6aea">

#### Proof
This confused me for a little while:

```ts
const num = parseInt('171b8fdad0', 16) // 17 1B 8F DA D0 -> 17 in hex === 23
(num >> 32) & 0xff // 208
(BigInt(num) >> 32n) & 0xffn // 23n
```

Reply:
I used the object id from your example that had 22 characters, if we use 24, we just need to adapt the code to:

```js
let temp = parseInt(hex.slice(0, 8), 16);

buffer[3] = temp & 0xff;
buffer[2] = (temp >> 8) & 0xff;
buffer[1] = (temp >> 16) & 0xff;
buffer[0] = (temp >> 24) & 0xff;

temp = parseInt(hex.slice(8, 16), 16);

buffer[7] = temp & 0xff;
buffer[6] = (temp >> 8) & 0xff;
buffer[5] = (temp >> 16) & 0xff;
buffer[4] = (temp >> 24) & 0xff;

temp = parseInt(hex.slice(16, 24), 16);

buffer[11] = temp & 0xff;
buffer[10] = (temp >> 8) & 0xff;
buffer[9] = (temp >> 16) & 0xff;
buffer[8] = (temp >> 24) & 0xff;
```

At least on this initial test, is working fine:

![image](https://github.com/mongodb/js-bson/assets/12551007/2a42afe9-3de5-4d78-943a-64550dcd5ff1)

> Number((BigInt(temp) >> 32n) & 0xffn);

I don't think the size of the string matters, we just need to parse the group of two characters from hex string in the correct order to be able to parse without issues.

Reply:
@H4ad You are completely right. My mistake. I've committed your code above except for some reason I was getting slightly better perf with `substring` instead of `slice`. 

What I like better about this implementation is we don't have to shim anything for browsers :+1: